### PR TITLE
DocC Documentation and Articles

### PR DESCRIPTION
**Pull request by hansfilbert**
# *DocC Documentation and Supporting Articles*

## :recycle: Current situation & Problem
Existing Documentation is structured in a folder

## :bulb: Proposed solution
Migrate the documentation to DocC with structured Documentation Catalog.

## :gear: Release Notes 
- Added Apodini documentation landing page
- Added Installation guide
- Added minimal Apodini example project
- Added extension files for multiple Protocols
- Added database connection guide
- No changes made on the existing documentation folder

## :heavy_plus_sign: Additional Information
Use Xcode 13 to build documentation.

### Related PRs
No related PR

### Testing
Tests are not included, irrelevant for documentation

### Reviewer Nudging
Check the compliance of Swift 5.5. This documentation requires Swift 5.5 to build
Apodini.docc folder under Source > Apodini

### CODE REVIEW

File: .github/workflows/build-and-test.yml
Comment (PSchmiedmayer) [inline .github/workflows/build-and-test.yml:24]: Please make sure that the GitHub Actions are exactly the same as on develop. Then all CI steps should pass as there are no changes to the source code itself 👍

File: Sources/Apodini/Apodini.docc/Tutorials/Apodini Installation.tutorial
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Tutorials/Apodini Installation.tutorial]: I think we can remove the tutorials as the articles already cover the content in a nice way.

**Comment by theMomax** [comment]
I just updated the [Delegating Handlers Wiki](https://github.com/Apodini/Apodini/wiki/Delegating-Handlers), because `Delegate.callAsFunction()` has been renamed to `Delegate.instance()` recently. @hansfilbert could you please update the docs here accordingly?

@Supereg I also noticed there is some outdated information on the parameters of `.delegated()`. Could you please also update the `Delegating-Handlers` page to reflect the recent changes you've made?

File: Sources/Apodini/Apodini.docc/Apodini.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Apodini.md]: This somehow seems off place here.

File: Sources/Apodini/Apodini.docc/AdditionalFunctionality/ProtocolBuffers.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/AdditionalFunctionality/ProtocolBuffers.md]: I would move this to the exporters group. This is related to the gRPC exporter.

File: Sources/Apodini/Apodini.docc/Extensions/Parameter.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Extensions/Parameter.md:1]: I would move this in the same group as the other property wrappers. Should me merged with the other `Parameter.md`

File: Sources/Apodini/Apodini.docc/Extensions/PathParameter.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Extensions/PathParameter.md:1]: Should be moved to the same group as `@Environment`

File: Sources/Apodini/Apodini.docc/GettingStarted/DatabaseConnection.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/GettingStarted/DatabaseConnection.md]: Maybe it makes sense to move this to additional functionality and refer to the articles there from the getting started section?

File: Sources/Apodini/Apodini.docc/Relationship/RetrieveRelationship.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Relationship/RetrieveRelationship.md]: I would put this in the Exporter Documentation category.

File: Sources/Apodini/Apodini.docc/Testing/NegativeCompileTest.md
Comment (PSchmiedmayer) [inline Sources/Apodini/Apodini.docc/Testing/NegativeCompileTest.md]: @Supereg What do you think: Should this be moved to the DocC documentation or should this stay somewhere else in the Repo for Apodini Maintainers? As far as I can see this is only usable internally and not an external feature.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
Hi @hansfilbert ,

thanks for updating the docs and moving them over to DocC. A really nice step forward! 🚀

I have a few general comments that I would liked to see addressed before merging
- Feel free to remove the tutorials, we already have the articles and I think they already cover the topic quite well
- Please move the REUSE comments right below the first `#` header line if this doesn't break the DocC generation
- Please remove all documentation that has been transferred to the DocC documentation from the documentation folder so we know what is left over there and still needs to be moved or dropped in the future
- I am not sure if "Tooling" is the correct summary for the `@Environment` property wrapper and so on. I would group them under `Component & Handler Properties` and would also group `@Parameter` and `@PathParameter` in the same group and reference this from the DSL components group

I have added a few additional comments here and there too.

In addition please choose two members of the @Apodini/apodini development team as reviewers for the PR to check the articles in detail and provide more detailed feedback on the content.

File: Sources/Apodini/Apodini.docc/Testing/NegativeCompileTest.md
Comment (bauer-andreas) [reply Sources/Apodini/Apodini.docc/Testing/NegativeCompileTest.md]: NegativeCompileTests are currently only used internally. Since it will only ever scan for tests targets in the Apodini repo at the moment, they can only be used internally.
So, at the current state, I don't think it makes sense to include them in the DocC documentation(?)

**Comment by theMomax** [comment]
There seems to be an issue with the DocC syntax here:

image

I also noticed there is hardly any code-sample in our docs that isn't outdated😐.

I think we should collaborate on the short term task of bringing the docs up to date and not leave it to @hansfilbert alone. We could either use this PR, or merge it first and start a second one. What strategy would you prefer?

In order to make sure the docs don't get this outdated again I'd propose three workflow changes:
* a DocC compilation job that fails on warnings could potentially capture outdated type-names
* we add section to the PR template to specify "Effects on Documentation"
* We try to keep the docs as close to the code as possible. I think DocC allows to separate type-documentation into two parts. The basic in-code documentation and an optional markdown based extension document. Is that correct @hansfilbert ? If so, I'd suggest to not use markdown extension documents at all for type-documentation, but keep everything as close to the swift code as possible. @hansfilbert would this restrict us in usage of any advanced DocC features that are only available in markdown files? The only thing that would be written in markdown would be tutorial and introduction documents.

**Comment by theMomax** [comment]
@hansfilbert I just pushed my latest version. This is should contain everything I intended to add/modify now. I restructured everything to get more of a tutorial/guide like experience similar to Vapor docs and wrote some new articles that focus on showing the usage to new users. I left some TODOs where we should link to future articles, or to external Articles (this is not possible with DocC right now, but it was requested quite often, so I assume it won't be long once DocC has been open-sourced).

@philippzagar It would be great if you could add an article regarding the USAGE of configuration with Swift Argument Parser (and also about Logging and Metrics once that is complete).

@Supereg It would also be great if you could add some small articles regarding metadata, information and authentication which really focus on usage. As far as I remember the inline-docs were already quite extensive in most these areas, so I think it will be enough to just copy-paste/link things together!

**Comment by PSchmiedmayer** [comment]
@theMomax Thanks a lot for the nice additions!

@hansfilbert Please also make sure to address the comments above, e.g. [the one about the GitHub Action file](https://github.com/Apodini/Apodini/pull/325#discussion_r679428211) to get all the checks green before we merge the PR 👍

@Supereg @philippzagar Please let @hansfilbert know if you want to add the documentation with this PR or open an other PR once you are done with your respective feature and/or want to create a small PR to update the documentation.

@hendesi I would also put you in the loop here. I think it would be good to add some detailed documentation about Apodini Deploy to the DocC documentation. As noted above this doesn't need to happen in this PR but It would be great if you can use the experience you have gained in the last weeks and months about the Deployment Providers as well as the changes you made in the deployment providers to highlight what Apodini Deploy and the concrete Deployment providers are capable of 🚀

**Comment by philippzagar** [comment]
@hansfilbert I already wrote a short article for [CLI argument parsing](https://github.com/Apodini/Apodini/wiki/Parse-CLI-arguments-in-the-WebService). Maybe you can use it as a template and basically copy-paste that into a DocC article. The benefit here is that the content would already be included in this PR. Otherwise, I'll do it myself later in a separate PR.

**Review by PSchmiedmayer — APPROVED** [review]

**Pull request merged**
